src/viam/sdk/resource/resource_base.hpp

```diff
@@ -20,12 +20,30 @@ class Resource {
     virtual ~Resource();
     explicit Resource(std::string name) : name_(std::move(name)){};
     static Subtype subtype();
+
+    /// @brief Returns the `Subtype` associated with a particular resource.
+    virtual Subtype instance_subtype() const = 0;
```

Comment:
This was necessary for subtype comparison. Previously calling to `subtype()` would return the base level `Resource` subtype unless the class in question happened to remember to implement a new static `Subtype()` method. This forces us to provide more granular access to a resource's subtype.

Reply:
Could we make `subtype()` virtual? It is a bit confusing to have two `subtype()` functions, especially when one calls the other

Reply:
We cannot, because `subtype()` is `static`. Unfortunately we have both the following cases:

- I know what resource type I want a `subtype` for, but I may not have access to an actual resource at the moment
- I have some resource but I don't know which kind it is, I would like to look at its subtype and compare

The static `subtype()` solves the first case, the virtual `instance_subtype()` solves the second. I would have loved to use the same name for these but I don't think you can overload a static method with a virtual one, at least when they both have the same signature.

Reply:
Ahh I see. I had missed the `static` on `subtype`...

Reply:
You are correct, you cannot overload on static vs virtual.

Could you provide a little more background on what the 'subtype' system is about? I've been meaning to ask anyway, and I'd like to make sure I understand it before I think about ways you could simplify this.

Reply:
Also a clarification on`Subtype` vs `ResourceSubtype` vs `ResourceType`.

Reply:
Yeah, so `Subtype`/`Type` as names are unfortunately a bit overloaded here. The short version is `Subtype` is the equivalent of RDK's `API` in `resource/api.go`. Basically it identifies a resource by its namespace (e.g. RDK), type (component/service), and subtype (camera/SLAM/motor). The name was changed recently from `Subtype` to `API`: I had been echoing the naming conventions in RDK and Python, and should probably continue coing so by renaming these types in a future PR.

`ResourceType` is just a string wrapper around a particular resource type, e.g. Camera or Motor.

`ResourceSubtype` is a collection of methods specific to a particular resource that get registered so we know how to, e.g., create a `CameraClient`. It is very poorly named, I wish I had a good answer for how I ended up with that name but I'm not sure 😅. Should definitely be renamed to `ResourceRegistration`.

Would prefer to not muddy the waters here by doing all the renaming, but a ticket is [here](https://viam.atlassian.net/browse/RSDK-2984).

Reply:
Yeah, definitely don't do the renaming in this review, but thanks for the clarification. Naming is hard!

Maybe call these `static_subtype` and `dynamic_subtype`. And then add `static_subtype` to `Resource`, `Component`, and `Service` so that it is consistently available throughout the hierarchy.

Another way forward would be to leave the dynamic version as `subtype` in `Resource`, and then use a template and specialization to make a `static Subtype Subtype::for<T>()` function where each T provides the result:


// In `subtype.hpp`:
```
class Subtype {
public:
    ....

    // Specialize subtype_map for each type T for which
    // Subtype::for<T>() should be defined. The specialization
    // should declare `static Subtype type()`.
    template<typename T>
    struct subtype_map;

    template<typename T>
    static Subtype for() {
        return subtype_map<T>::type();
    };
};
```

In `camera.hpp`:
```
class Camera {
public:
    ....
    /// @override Resource::subtype
    Subtype subtype() const override { return SubType::for<Camera>() };
    ...
};

template<>
struct Subtype::subtype_map<Camera> {
    static Subtype type() {
        return {RDK, COMPONENT, "camera"};
    }
};
```

Obviously, the actual implementations do not need to be inline / declared in the class body. Basically, this gets rid of the collision between the static / virtual by making the way to get the static subtype of something part of the `Subtype` API rather than the per-type API.

Usage now looks like:

```
// Get the runtime subtype of an object via its `Resource` pointer (or anywhere else in class hierarchy).
auto subtype = r->subtype(); 

// Get the static subtype of a thing:
auto subtype = Subtype::for<Camera>();
```

Just some (untested) ideas!

Reply:
Oh, I do like the look of `Subtype::for<T>();`! That looks fun and clean. For the purpose of this PR though I think I'm going to go with the low hanging fruit and just rename to `{static|dynamic}_subtype()`. Ticket for looking into better, more involved solutions [here](https://viam.atlassian.net/browse/RSDK-2990).